File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
from typing import Optional

class Solution:
    def lowestCommonAncestor(self, root: 'TreeNode', p: 'TreeNode', q: 'TreeNode') -> 'TreeNode':
        if root is None:
            return None

        if root == p or root == q:
            return root

        lca_left = self.lowestCommonAncestor(root.left, p, q)
        lca_right = self.lowestCommonAncestor(root.right, p, q)

        if lca_left and lca_right:
            return root

        return lca_left if lca_left else lca_right

File: 0236-lowest-common-ancestor-of-a-binary-tree/0236-lowest-common-ancestor-of-a-binary-tree.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function lowestCommonAncestor(root: TreeNode | null, p: TreeNode | null, q: TreeNode | null): TreeNode | null {
    if (root === null) {
        return null;
    }

    if (root === p || root === q) {
        return root;
    }

    const lcaLeft = lowestCommonAncestor(root.left, p, q);
    const lcaRight = lowestCommonAncestor(root.right, p, q);

    if (lcaLeft && lcaRight) {
        return root;
    }

    return lcaLeft ?? lcaRight;
}	
